### PR TITLE
add a delimiters property to override enter/tab

### PR DESCRIPTION
**Pull request by Pomax**
This change adds custom delimiters that, when used, *bypasses* the ENTER and TAB keys for committing a new tag. This restores default form behaviour for the tab key ("move to next form element"), and makes the enter key a noop, instead allowing the user to specify that new tags should be committed based on commas, spaces, etc. to match the UX similar to how several major sites handle tags (e.g. Stackoverflow, Flickr, etc.)

This PR closes https://github.com/i-like-robots/react-tags/issues/76

### CODE REVIEW

File: lib/ReactTags.js [hunk 1]
```diff
@@ -60,7 +60,14 @@ class ReactTags extends React.Component {
     const { query, selectedIndex } = this.state
 
     // when one of the terminating keys is pressed, add current query to the tags.
-    if (e.keyCode === KEYS.ENTER || e.keyCode === KEYS.TAB) {
+    let shouldAdd = e.keyCode === KEYS.ENTER || e.keyCode === KEYS.TAB
+
+    // if delimiters were explicitly provided, reasses based on delimiters.
+    if (this.props.delimiters) {
```
Comment (markus101) [inline lib/ReactTags.js]: Personally I'd set `ReactTags.defaultProps.delimiters = [KEYS.ENTER, KEYS.TAB]` instead, then you're only checking one place. Probably worthwhile to verify that at least one item is present as well.

File: lib/ReactTags.js [hunk 2]
```diff
@@ -203,6 +210,7 @@ ReactTags.propTypes = {
   suggestions: PropTypes.array,
   autofocus: PropTypes.bool,
   autoresize: PropTypes.bool,
+  delimiters: PropTypes.array,
```
Comment (markus101) [inline lib/ReactTags.js]: With a default value I'd make this `isRequired` as well.

File: lib/ReactTags.js (hunk 1 above)
Comment (Pomax) [inline lib/ReactTags.js]: good point

**Comment by Pomax** [comment]
@markus101 updated, although it does require people provide delimiters as `<ReactTags delimiters={[ ','.charCodeAt(0) ]} ......>`

That's not terrible, but probably needs README updates

**Comment by markus101** [comment]
The only thing that would need to be checked is whether delimiters is empty (if you set to an empty array, maybe that's on the developer), otherwise I think it's fine.

**Comment by Pomax** [comment]
@markus101 changed it over to a custom PropType validator so that we can test that it's (a) and array and (b) not empty

File: lib/ReactTags.js [hunk 3]
```diff
@@ -203,6 +204,25 @@ ReactTags.propTypes = {
   suggestions: PropTypes.array,
   autofocus: PropTypes.bool,
   autoresize: PropTypes.bool,
+  delimiters: function (props, propName, componentName) {
```
Comment (markus101) [inline lib/ReactTags.js]: Nice, it's too bad it needa to do the array validation, I was going to suggest :

```
  // You can also supply a custom validator to `arrayOf` and `objectOf`.
  // It should return an Error object if the validation fails. The validator
  // will be called for each key in the array or object. The first two
  // arguments of the validator are the array or object itself, and the
  // current item's key.
  customArrayProp: PropTypes.arrayOf(function(propValue, key, componentName, location, propFullName) {
    if (!/matchme/.test(propValue[key])) {
      return new Error(
        'Invalid prop `' + propFullName + '` supplied to' +
        ' `' + componentName + '`. Validation failed.'
      );
    }
  })
```

(from https://facebook.github.io/react/docs/typechecking-with-proptypes.html), but if the array is empty the custom validation isn't triggered.

Comment (Pomax) [inline lib/ReactTags.js]: yeah, I was hoping to keep the array proptype in somehow, but if the idea was to test for an empty array the entire thing needs validating =(

**Comment by Pomax** [comment]
Not sure how to fix the coveralls issue - is there a way to see what part of the file has lost coverage?

**Comment by Pomax** [comment]
b00g confirmed. time to have a look at what the forward function is doing in the `is_approved` → `moderation state` migration

**Comment by Pomax** [comment]
@markus101 any idea what coverage was lost?

**Comment by markus101** [comment]
Perhaps related to the custom validation? That's the big change in the last two commits and my guess is the tester ran the same SHA twice because they were so close or used the last SHA for the messaging.

A couple specs to ensure delimiters is an array and isn't empty can't be empty would help.

**Comment by Pomax** [comment]
not sure where to put that, normally that's what PropTypes does.

**Comment by markus101** [comment]
https://stackoverflow.com/a/31835256/882971 - is what I was thinking (not great that it'd need to spy on `console.error` though). Perhaps @i-like-robots has a suggestion.

**Comment by markus101** [comment]
Actually after some more thought it'd probably be best to extract the validator to another file and add specs to test the function directly.

**Comment by Pomax** [comment]
yeah that makes sense. let me add that in some time today.

**Comment by Pomax** [comment]
@markus101 refactored the validator to its own file, with tests added to the spec file to ensure it's accepting and rejecting the correct data.

**Comment by Pomax** [comment]
how is the README.md maintained, btw? I can add a section on the delimiters in the options, but there's some weird code in the README for navigation that looks like it could both be handwritten and auto-generated (HTML4 `<a name>` is used for anchoring, something that was taken out of HTML5, where anchoring is achieved through `id` attributes, and `<a>` are only ever true links to other resources).

File: lib/DelimiterValidator.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  // and that it's not an empty array
+  if (prop.some(e => typeof e !== 'number')) {
+    return new Error(
+      'Invalid prop `' + propName + '` supplied to' +
+      ' `' + componentName + '`: delimiters must be keyCode integers.'
+    )
+  }
+}
```
Comment (i-like-robots) [inline lib/DelimiterValidator.js]: This is interesting, I'd not seen custom validators in use before. I wonder if this can be compiled out easily or not?

**Comment by i-like-robots** [comment]
Thanks for putting this together @Pomax, it looks good.

Re. the readme it's all done manually, I've just been maintaining the style of the original project.

**Comment by Pomax** [comment]
@i-like-robots gotcha. Mind if I change those `<a name="..."></a>` into `<span id="..."></span>` instead? (to make it a little more HTML5-semantically conformant?)

**Comment by i-like-robots** [comment]
@Pomax feel free! Seeing anchors used in such a way makes me feel nostalgic 👴

**Comment by i-like-robots** [comment]
Also don't worry about coveralls, when I merge this in I'll sort it out before releasing

**Comment by Pomax** [comment]
@i-like-robots alright, README.md updated, as near as I can figure this is done =)

**Comment by i-like-robots** [comment]
Fabulous, thank you!

**Comment by Pomax** [comment]
looking forward to a minor release update so we can tap into this =D

**Pull request merged**